---
title: "Analyzing pooled CRISPR functional genetic screens with MAGeCKFlute"
author: "WubingZhang, Feizhen Wu, and Binbin Wang"
date: "24 Oct 2017"
package: "0.99.9"
abstract: >
  The purpose to do CRISPR screen is to filter essential genes, pathways 
  and to explain the background mechanisms. We developed MAGeCKFlute to
  perform integrated analysis of CRISPR/Cas9 screens with/without drug
  treatments. The MAGeCKFlute provides several strategies to remove 
  potential biases within read counts and beta scores. The downstream
  analysis for CBS(Control Beta Score) and TBS(Treatment Beta Score) 
  with the package includes identifying essential, non- essential, and
  drug-associated genes, and performing biological functional analysis for
  these genes. The package also visualizes genes in the context of pathways
  to better help users explore the screening data. Collectively, MAGeCKFlute
  enables accurate identification of essential, non-essential, drug-targeted
  genes, as well as their related biological functions. This vignette
  explains the use of the package and demonstrates typical workflows.
 
  MAGeCKFlute package version: `r packageVersion("MAGeCKFlute")`
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{MAGeCKFlute.Rmd}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


<!-- This is the source document -->

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=FALSE, error=FALSE, warning=TRUE)
```	

## Quick start

Here we show the most basic steps for an analysis pipeline of pooled
CRISPR functional genetic screens. There are a variety of steps
upstream of MAGeCKFlute that result in the generation of FDRs of
positive selection and negative selection, or beta scores for each gene
in all conditions, which we will discuss in the sections below. 

To run MAGeCKFlute pipeline, we need gene summary file generated by
running MAGeCK. MAGeCKFlute package provides two example data, one is
`BRAF_mle.gene_sunmmary` and the other is `BRAF_rra.gene_summary`. 
We will work with them in this document.

**Run pipeline from gene summary file generated by MAGeCK MLE**
```{r quickStart1, eval=FALSE}
##Load MLE gene summary data
data("BRAF_mle.gene_summary")
##Run the MAGeCKFlute MLE pipeline
FluteMLE(gene_summary=BRAF_mle.gene_summary, 
         ctrlname=c("D7_R1", "D7_R2"), 
         treatname=c("PLX7_R1","PLX7_R2"), 
         prefix="BRAF", organism="hsa")

#All pipeline results are written into local directory 
# "./BRAF_Flute_Results/", and all figures are integrated into file
# "BRAF_Flute.mle_summary.pdf".
```


**Run pipeline from gene summary file generated by MAGeCK RRA**

```{r quickStart2, eval=FALSE}
##Load RRA gene summary
data("BRAF_rra.gene_summary")
##Run the MAGeCKFlute RRA pipeline
FluteRRA(BRAF_rra.gene_summary, prefix="BRAF", organism="hsa")

#All pipeline results are written into local directory
# "./BRAF_Flute_Results/" too, and all figures are integrated into file 
# "BRAF_Flute.rra_summary.pdf".
```


## How to get help for MAGeCKFlute

Any and all MAGeCKFlute questions should be posted to the 
**Bioconductor support site**, which serves as a searchable knowledge
base of questions and answers:

<https://support.bioconductor.org>

Posting a question and tagging with "MAGeCKFlute" will automatically send
an alert to the package authors to respond on the support site.  See
the first question in the list of [Frequently Asked Questions](#FAQ)
(FAQ) for information about how to construct an informative post. 

You should **not** email your question to the package authors, as we will
just reply that the question should be posted to the 
**Bioconductor support site**.


## Input data

As input, the MAGeCKFlute package expects gene summary file as obtained
by running commands `mageck test` or `mageck mle` in MAGeCK and
MAGeCK-VISPR, which are developed by our lab previously, to analyze
CRISPR/Cas9 screen data in different scenarios. Both algorithms use
negative binomial models to model the variances of sgRNAs, and use Robust
Rank Aggregation (for MAGeCK) or maximum likelihood framework (for
MAGeCK-VISPR) for a robust identification of selected genes.

Samples of CRISPR screen with drug-treatment are generally designed as
three components: initial sample (usually the *day0 sample*), *drug
treated sample* and corresponding *controls* (usually with vehicle
treatment). The *day0* sample represents cells in the start time when
screening library was transferred into cells, and the *treatments* and
*controls* are samples treated with or without drug, respectively.
MAGeCK-VISPR uses a negative-binomial likelihood model to assess the
functions of genes in a screen, and outputs a **gene summary file**, in
which *control beta scores (CBS)* and *treatment beta scores (TBS)* are
calculated by comparing the control and treatment samples to day0,
respectively. The `beta score` for each gene describes how the gene is
selected: a positive beta score indicates a positive selection (and vice
versa), similar to the value of log2-transformed fold-change in gene
differential expression analysis. The format of **gene summary file**
outputted by MAGeCK-VISPR is same as that of `BRAF_mle.gene_sunmmary`.

CRISPR screens without drug-treatment only contain two kind of samples,
*day0 sample* and *control sample*. Then in this case, you can use MAGeCK
RRA to identify negatively and positively selected genes. MAGeCK compares
two experimental condition samples, performs a statistical test for each
gene and caculates FDRs for positive and negative selection, which are
summarized into **gene summary file**, similar to
`BRAF_rra.gene_sunmmary`.

In a word, before using MAGeCKFlute package and running pipeline, you
need a gene summary file like `BRAF_mle.gene_sunmmary` or
`BRAF_rra.gene_sunmmary`.


## Section I: Run pipeline from gene summary file generated by MAGeCK MLE

### Gene summary file 

As the input of MAGeCKFlute package, the gene summary file from MAGeCK-MLE
includes beta scores of all genes in control and treatment samples. The
"beta score" is similar to the value of log2-transformed fold-change in
differential expression analysis. Use function ‘data’ to load the dataset,
and have a look at the file with a text editor to see how it is formatted.

```{r CheckMLERes}
library(MAGeCKFlute)
data("BRAF_mle.gene_summary")
head(BRAF_mle.gene_summary)
```


### Read beta scores

Then, extract beta scores of control and treatment samples from the
gene summary data matrix, and  have a look at the beta score matrix.

```{r ReadBeta}
gene_summary=BRAF_mle.gene_summary
ctrlName=c("D7_R1", "D7_R2")
treatName=c("PLX7_R1", "PLX7_R2")
#Read beta scores from gene summary matrix
dd=ReadBeta(gene_summary, ctrlName=ctrlName, treatName=treatName, organism="hsa")
head(dd)
```


### Normalization of beta scores

It is difficult to control all samples with a consistent cell cycle in an
CRISPR screen experiment with multi conditions. Besides, beta score among
different conditions with an inconsistent cell cycle are incomparable. So
it is necessary to do the normalization when comparing the beta scores in
different conditions. Essential genesare thosegenes that are indispensable
for its survival. The effect generated by knocking out these genes in
different cell types is consistent. Based on this, we developed cell cycle
normalization method to shorten the gap of cell cycle in different
conditions. In addition, a previous normalization method called loess
normalization is available in this package. These two normalization methods
are optional.

```{r NormalizeBeta}
dd_essential = NormalizeBeta(dd, method="cell_cycle")
head(dd_essential)

#OR
dd_loess = NormalizeBeta(dd, method="loess")
head(dd_loess)
```


### Distribution of all gene beta scores

After normalization, the distribution of beta scores in different
conditions should be similar. We can evaluate the distribution of beta
scores using function ‘Violin.plot’, ‘Density.plot’, and ‘Density.diff’.

```{r DistributeBeta}
Violin.plot(dd_essential, main="Cell cycle normalized")
Density.plot(dd_essential, main="Cell cycle normalized")
Density.diff(dd_essential, main="Cell cycle normalized")

#MAplot was used to evaluate the data quality of gene expression profile 
#in the past, here we can use it to evaluate the data quality of normalized
#beta score profile.
MA.plot(dd_essential, cex=1, main="Cell cycle normalized")
```

### Estimate cell cycle time by linear fitting

After normalization, the cell cycle time in different condition should be almost consistent. Here we use linear fitting to estimate the cell cycle time in different samples with the sample ‘Control’ as x- axis control.

```{r EstimateCellCycle}
##Fitting beta score of all genes
CellCycleFit(dd_essential, ylab="Beta Score", main="Cell cycle normalized")
```

### Positive selection and negative selection

We can use function ‘GroupAB’ to group all genes into three groups,
positive selection genes (GroupA), negative selection genes (GroupB), 
and others. Then, visualize these three grouped genes in scatter plot 
using function ‘ScatterAB’ or use function ‘RankAB’ to rank the beta score
deviation between control and treatment and mark top selected genes in
figure.

```{r selection}
ddAB_essential=GroupAB(dd_essential)
ScatterAB(ddAB_essential, main="Cell cycle normalized")
RankAB(ddAB_essential, main="Cell cycle normalized")
```

### Functional analysis of selected genes

Gene ontology-biological process and Kyoto encyclopedia of genes and
genomes (KEGG) pathway enrichment analysis are available in this pipeline.
For GroupA and GroupB genes, we use function ‘EnrichAB’ to do all useful
enrichment analysis.

```{r EnrichAB, eval=FALSE}
enrich_result = EnrichAB(ddAB_essential,pvalue=0.05, 
                         enrich_method = "Hypergeometric", organism="hsa")
print(enrich_result$keggA$gridPlot)
print(enrich_result$keggB$gridPlot)
```

For enriched pathways, we can use function "KeggPathwayView" to visualize
the beta score level in control and treatment on pathway map.

```{r pathview, eval=FALSE,eval=FALSE}
genedata = dd_essential[,c("Control","Treatment")]
rownames(genedata)=dd_essential$ENTREZID
keggID = enrich_result$keggA$enrichRes@result$ID[1]
KeggPathwayView(gene.data = genedata, pathway.id = keggID, species="hsa")
#The pathway map will be located on current workspace
```

### Filter essential genes in specific condition using 9-square model

Considering the difference of beta scores in control and treatment sample,
we developed a 9-square model, which group all genes into several
subgroups. Among these subgroups, four subgroup genes are essential, which
correspond to specific functions. Group1 and Group3 genes are not selected
in control sample, while they are significantly selected in treatment
sample, so these genes may related to drug resistance. Group2 and Group4
genes are selected in control, but they are not selected in treatment, so
maybe these genes are associated with drug targets.
In this package, use function ‘Square.plot’ can select these four
meaningful subgroup genes and get the 9-Square scatter plot object.

```{r 9Square}
dd2=dd_essential[,c("Gene","Treatment","Control","ENTREZID")]
P2=Square.plot(dd2,main="Cell cycle normalized")
print(P2$p)
```


### Functional analysis of essential genes in specific condition

For all the four subgroup genes and combination of them, BP and KEGG
enrichment analysis can be performed easily using ‘EnrichSquare’ function.

```{r EnrichSquare,eval=FALSE}
enrich_result2 = EnrichSquare(P2$dd1,pvalue=0.05)
print(enrich_result2$kegg3$gridPlot)
print(enrich_result2$kegg4$gridPlot)
```

Pathway visualization can be done using function ‘KeggPathwayView’, the same as section above.

```{r pathview2, eval=FALSE,eval=FALSE}
genedata = dd_essential[,c("Control","Treatment")]
rownames(genedata)=dd_essential$ENTREZID
keggID = enrich_result2$kegg1$enrichRes@result$ID[1]
KeggPathwayView(gene.data = genedata, pathway.id = keggID, species="hsa")
```

## Section II: Run pipeline from gene summary file generated by MAGeCK RRA

### Gene summary file
For experiments without, we recommend to use MAGeCK-RRA to identify essential genes from CRISPR/Cas9 knockout screens and tests the statistical significance of each observed change between two conditions. We use the term 'essential' to refer to positively or negatively selected genes. Gene summary file in MAGeCK-RRA results summarize the statistical significance of positive selection and negative selection.

```{r CheckRRARes}
data("BRAF_rra.gene_summary")
head(BRAF_rra.gene_summary)
```

### Read FDR of negative selection and positive selection

Then, extract "neg.fdr" and "pos.fdr" from the gene summary data matrix,
and  have a look at the matrix.

```{r ReadRRA}
dd.rra = ReadRRA(BRAF_rra.gene_summary, organism="hsa")
head(dd.rra)
```

### Negative selection and positive selection

Take 0.05 as cutoff, get negative selection and positive selection genes and do enrichment analysis on KEGG pathway and GO BP terms.

```{r selection2, eval=FALSE}
##Negative selection
universe=dd.rra$ENTREZID
genes = dd.rra[dd.rra$neg.fdr<0.05, "ENTREZID"]
kegg.neg=enrichment_analysis(geneList = genes, universe=universe, 
                             type = "KEGG", method="HyperGeometric", 
                             pvalueCutoff = 0.05, plotTitle="KEGG: neg")
bp.neg=enrichment_analysis(geneList = genes, universe=universe, 
                           method = "ORT", type = "BP", 
                           pvalueCutoff = 0.05, plotTitle="BP: neg")
print(kegg.neg$gridPlot)
print(bp.neg$gridPlot)

##Positive selection
universe=dd.rra$ENTREZID
genes = dd.rra[dd.rra$pos.fdr<0.05, "ENTREZID"]
kegg.pos=enrichment_analysis(geneList = genes, universe=universe, 
                             type = "KEGG", method="HyperGeometric", 
                             pvalueCutoff = 0.05, plotTitle="KEGG: pos")
bp.pos=enrichment_analysis(geneList = genes, universe=universe, 
                           method = "ORT", type = "BP", 
                           pvalueCutoff = 0.05, plotTitle="BP: pos")
print(kegg.pos$gridPlot)
print(bp.pos$gridPlot)

```



# Session info

```{r sessionInfo}
sessionInfo()
```

# References

